server/controllers/profile/getInternProfile.js

```diff
@@ -1,29 +1,87 @@
 
 const { getUserReviews } = require("./../../database/queries/user/index");
+const { getInternBookingsWithReviews } = require("./../../database/queries/bookings");
 
 module.exports = async (req, res, next) => {
-  const { user: { _id: userId } } = req;
+  try {
+    const { id: internId } = req.params;
+    const { _id: userId, role } = req.user;
+    const { expand } = req.query;
 
-  // check if user id is in request
-  if (!userId) {
-    return next(boom.badRequest("error loading profile"));
-  }
-  const profileData = id => Promise.all([internProfileData(id), getUserReviews(id)]);
+    // check if user id is in request
+    if (!internId) {
+      return next(boom.badRequest("error loading profile"));
+    }
+
+    if (!mongoose.Types.ObjectId.isValid(internId)) {
+      return next(boom.notFound());
+    }
+
+    if (internId !== userId.toString() && role !== "admin") {
+      return next(boom.unauthorized());
+    }
+
+    // Basic profile
+    const dataParams = [{ userInfo: "single" }];
+    const promisesArray = [internProfileData(internId)];
+
+    if (expand && expand.includes("bookings") && expand.includes("reviews")) {
+    // get the intern's bookings and with reviews
+      dataParams.push({ bookingsWithReviews: "array" });
+      promisesArray.push(getInternBookingsWithReviews(internId));
+    } else if (expand && expand.includes("reviews")) {
+    // get the intern's reviews
+      dataParams.push({ reviews: "array" });
+      promisesArray.push(getUserReviews(internId));
+    }
+
+
+    const data = await Promise.all(promisesArray);
 
-  return profileData(userId)
-    .then((data) => {
-      // if no profile data return 404
-      if (!data[0].length) {
-        return next(boom.notFound("Cannot find the profile you're looking for"));
+    // if no profile data return 404
+    if (!data[0].length) {
+      return next(boom.notFound("Cannot find the profile you're looking for"));
+    }
+
+    const formedData = {};
+    dataParams.forEach((item, index) => {
+      const [key, value] = Object.entries(item)[0];
+
+      if (value === "single") {
+        // eslint-disable-next-line prefer-destructuring
+        formedData[key] = data[index][0];
+      } else {
+        formedData[key] = data[index];
+      }
+    });
+
+    const { profile } = formedData.userInfo;
+
+    // get the file links
+    if (profile) {
+      const promises = [];
+      if (profile.verification) {
+        promises.push(generateUrl(profile.verification.photoID));
+        promises.push(generateUrl(profile.verification.offerLetter));
+      }
+      if (profile.profileImage) {
+        promises.push(generateUrl(profile.profileImage));
       }
-      return res.json(data);
-    })
-    .catch(() => next(boom.badRequest("error loading profile")));
+      await Promise.all(promises);
+    }
+
+    return res.json(formedData);
+  } catch (error) {
+    return next(boom.badImplementation("error loading profile"));
```

Comment:
I think it is always good practice to send the error object to the error middleware
```suggestion
    return next(boom.badImplementation(error));
```